### PR TITLE
id-rails-console triggers Slack notification

### PR DESCRIPTION
**Pull request by bleachbyte**
Running `id-rails-console` from any host (or, in particular, an IDP host) will now trigger a Slack notification in the environment's appropriate `slackchannel` (e.g. `#login-events` for prod, `#login-otherevents` for everything else):

```
id-rails-consoleAPP 3:13 PM
:rails:rails console:rails: for ubuntu on idp-i-0811f09c6f10dfe7f.ci.login.gov
use case: "testing"
```

Fixes https://github.com/18F/identity-devops/issues/1550.

### CODE REVIEW

File: identity_base_config/recipes/default.rb [hunk 1]
```diff
@@ -64,19 +64,16 @@
...
-  group 'root'
-  mode '0755'
-end
+template "/usr/local/bin/id-rails-console" do
+    source "id-rails-console.erb"
+    owner 'root'
+    group 'root'
+    mode '0755'
+    variables ({
+      :webhook => ConfigLoader.load_config(node, "slackwebhook"),
```
Comment (brodygov) [inline identity_base_config/recipes/default.rb]: We should drop the credentials into a separate file, not a template, so we can make the perms non world readable while still having id-rails-console itself world readable.

File: identity_base_config/templates/default/id-rails-console.erb [hunk 2]
```diff
@@ -40,6 +43,20 @@ if [ $# -ge 1 ]; then
...
+    read -r -p "Please provide an explanation for this Rails console: " USE_CASE
+    if [ ! -z "$USE_CASE" ]; then
+        break
+    fi
+done
+
+CURL='curl -X POST --data-urlencode '
+TEXT=":rails:*rails console*:rails: for \`$(logname)\` on \`$(hostname)\`\nuse case: \\\"${USE_CASE}\\\""
+PAYLOAD="'payload={\"channel\": \"<%= @channel %>\", \"username\": \"id-rails-console\", \"text\": \"${TEXT}\", \"icon_emoji\": \":dopetopus:\"}'"
+bash -c "${CURL} ${PAYLOAD} <%= @webhook %>"
```
Comment (brodygov) [inline identity_base_config/templates/default/id-rails-console.erb]: Why do we need to open another bash?

File: identity_base_config/recipes/default.rb (hunk 1 above)
Comment (bleachbyte) [reply identity_base_config/recipes/default.rb]: Ah hah, that methodology is what I was missing; will correct.

File: identity_base_config/templates/default/id-rails-console.erb (hunk 2 above)
Comment (bleachbyte) [reply identity_base_config/templates/default/id-rails-console.erb]: This was based off of the additions in `deploy` here: https://github.com/18F/identity-devops/pull/119/files#diff-078f40fa23e0672777adc7c05d4773ddR55-R62 but I can knock it down to just the `curl` command with the `$PAYLOAD`.

File: identity_base_config/files/default/id-rails-console [hunk 3]
```diff
@@ -40,6 +44,24 @@ if [ $# -ge 1 ]; then
...
+CHANNEL=$(cat /etc/login.gov/keys/slackchannel) || ((missing_keys++))
+WEBHOOK=$(cat /etc/login.gov/keys/slackwebhook) || ((missing_keys++))
+if [ ${missing_keys} -gt 0 ]; then
+    echo 'Slack channel/webhook missing! Verify files and run Chef to repopulate.'
+    exit 1
+fi
+
+TEXT=":rails:*rails console*:rails: for \`$(logname)\` on \`$(hostname)\`\nuse case: \\\"${USE_CASE}\\\""
+PAYLOAD="'payload={\"channel\": \"${CHANNEL}\", \"username\": \"id-rails-console\", \"text\": \"${TEXT}\", \"icon_emoji\": \":dopetopus:\"}'"
+curl -X POST --data-urlencode "${PAYLOAD} ${WEBHOOK}"
```
Comment (brodygov) [inline identity_base_config/files/default/id-rails-console]: Shouldn't payload and webhook should be separate arguments?

File: identity_base_config/files/default/id-rails-console [hunk 4]
```diff
@@ -40,6 +44,24 @@ if [ $# -ge 1 ]; then
...
+done
+
+CHANNEL=$(cat /etc/login.gov/keys/slackchannel) || ((missing_keys++))
+WEBHOOK=$(cat /etc/login.gov/keys/slackwebhook) || ((missing_keys++))
+if [ ${missing_keys} -gt 0 ]; then
+    echo 'Slack channel/webhook missing! Verify files and run Chef to repopulate.'
+    exit 1
+fi
+
+TEXT=":rails:*rails console*:rails: for \`$(logname)\` on \`$(hostname)\`\nuse case: \\\"${USE_CASE}\\\""
```
Comment (brodygov) [inline identity_base_config/files/default/id-rails-console]: Do we need to worry about escaping characters in USE_CASE? What will happen if someone includes typical punctuation or newlines?

File: identity_base_config/recipes/default.rb [hunk 5]
```diff
@@ -64,11 +64,16 @@
...
-  owner 'root'
-  group 'root'
-  mode '0755'
+# download Slack hook/channel files for id-rails-console
+aws_account_id = AwsMetadata.get_aws_account_id
+
+s3_secret_url = "s3://login-gov.secrets.#{aws_account_id}-us-west-2/#{node.chef_environment}/"
+
+%w(slackwebhook slackchannel).each do |f|
+  execute "download #{f} from s3" do
```
Comment (brodygov) [inline identity_base_config/recipes/default.rb]: Whoa, these should be `file` resources, not `execute`.

https://github.com/18F/identity-devops/blob/9928e1ec250dfe55b9207686b1eca50bb01046e5/kitchen/cookbooks/config_loader/recipes/default.rb#L8-L10

File: identity_base_config/files/default/id-rails-console (hunk 3 above)
Comment (bleachbyte) [reply identity_base_config/files/default/id-rails-console]: I was thinking of the previous behavior that utilized a full string for the command; will correct.

File: identity_base_config/recipes/default.rb (hunk 5 above)
Comment (bleachbyte) [reply identity_base_config/recipes/default.rb]: I was basing this off the behavior we currently have in `identity-elk`, e.g.:

https://github.com/18F/identity-devops/blob/9928e1ec250dfe55b9207686b1eca50bb01046e5/kitchen/cookbooks/identity-elk/recipes/es.rb#L162-L174

File: identity_base_config/files/default/id-rails-console (hunk 4 above)
Comment (bleachbyte) [reply identity_base_config/files/default/id-rails-console]: I added (and tested) a line to parse and escape characters using `printf '%q'` and reassign that value to `$USE_CASE`:
```
:~$ USE_CASE="Testing this
> weird thing! What 'will' it do?"
:~$ printf '%q\n' "$USE_CASE"
$'Testing this\nweird thing! What \'will\' it do?'
:~$ USE_CASE=$(printf '%q\n' "${USE_CASE}" | sed 's/\$//')
:~$ echo $USE_CASE
'Testing this\nweird thing! What \'will\' it do?'
```

File: identity_base_config/recipes/default.rb [hunk 6]
```diff
@@ -64,11 +64,11 @@
   mode '0755'
 end
 
-cookbook_file '/usr/local/bin/id-git' do
```
Comment (brodygov) [inline identity_base_config/recipes/default.rb]: Don't think we mean to delete `id-git`

File: identity_base_config/recipes/default.rb [hunk 7]
```diff
@@ -64,11 +64,11 @@
...
-  source 'id-git'
-  owner 'root'
-  group 'root'
-  mode '0755'
+# download Slack hook/channel files for id-rails-console
+%w(slackwebhook slackchannel).each do |f|
+  file "/etc/login.gov/keys/#{f}" do
+    content ConfigLoader.load_config(node, "#{f}")
```
Comment (brodygov) [inline identity_base_config/recipes/default.rb]: Nit: `"#{f}"` can be `f` instead

File: identity_base_config/recipes/default.rb (hunk 6 above)
Comment (bleachbyte) [reply identity_base_config/recipes/default.rb]: That... is odd. Must have been a misfire; fixing.

**Comment by brodygov** [comment]
P.S. Please be sure to squash the history here.

**Comment by bleachbyte** [comment]
Using `bash -c` before the `curl` command makes it easier to put the full command together, but ultimately requires a LOT more quote-escapery. Once I removed all of that wizardry, and declared the strings using single quotes instead of double, `curl` worked by itself.

**Comment by brodygov** [comment]
Yeah, using `bash -c` from within bash is almost always a bad idea.

**Pull request merged**